### PR TITLE
[Module] managedClusters - Add ingressProfile support for AKS

### LINKED ISSUE #3183: [Feature Request]: managedClusters - Add ingressProfile support for AKS
### Description

According to the documentation for Microsoft.ContainerService managedClusters at https://learn.microsoft.com/en-us/azure/templates/microsoft.containerservice/managedclusters?pivots=deployment-language-bicep, the "IngressProfile" feature is currently supported in the managedClusters module. 

As a result, managedClusters module needs to include "ingressProfile" section:

```
ingressProfile: {
      webAppRouting: {
        dnsZoneResourceId: 'string'
        enabled: bool
      }
    }
```

This will require updating the API versions of the module to 2023-03-02-preview as well.

### PR DESCRIPTION
**Pull request by samuelya**
# Description

closes #3183

| Pipeline |
| - |
| [ContainerService - ManagedClusters](https://github.com/samuelya/ResourceModules/actions/workflows/ms.containerservice.managedclusters.yml)|

# Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: modules/Microsoft.ContainerService/managedClusters/main.bicep [hunk 1]
```diff
@@ -132,6 +132,12 @@ param agentPools array = []
 @description('Optional. Specifies whether the httpApplicationRouting add-on is enabled or not.')
 param httpApplicationRoutingEnabled bool = false
 
+@description('Optional. Specifies whether the webApplicationRoutingEnabled add-on is enabled or not.')
+param webApplicationRoutingEnabled bool = false
+
+@description('Optional. Specifies the resource ID of connected DNS zone. Optional if `webApplicationRoutingEnabled` is set to `true`.')
```
Comment (AlexanderSehr) [inline modules/Microsoft.ContainerService/managedClusters/main.bicep:138]:
```suggestion
@description('Conditional. Specifies the resource ID of connected DNS zone. Required if `webApplicationRoutingEnabled` is set to `false`.')
```

File: modules/Microsoft.ContainerService/managedClusters/main.bicep [hunk 2]
```diff
@@ -619,6 +631,20 @@ module managedCluster_roleAssignments '.bicep/nested_roleAssignments.bicep' = [f
   }
 }]
 
+resource dnsZone 'Microsoft.Network/dnsZones@2018-05-01' existing = if (dnsZoneResourceId != null && webApplicationRoutingEnabled) {
+  name: last(split(dnsZoneResourceId, '/'))!
```
Comment (AlexanderSehr) [inline modules/Microsoft.ContainerService/managedClusters/main.bicep:635]: I guess there is a chance that the DNS Zone is in a different subscription / resource Group? In this case I'd suggest as smal update:

```suggestion
  name: last(split(dnsZoneResourceId, '/'))!
  scope: resourceGroup(split(dnsZoneResourceId, '/')[2], split(dnsZoneResourceId, '/')[4])
```

File: modules/Microsoft.DocumentDB/databaseAccounts/readme.md [hunk 3]
```diff
@@ -39,6 +39,7 @@ This module deploys a DocumentDB database account and its child resources.
 | Parameter Name | Type | Default Value | Allowed Values | Description |
 | :-- | :-- | :-- | :-- | :-- |
 | `automaticFailover` | bool | `True` |  | Enable automatic failover for regions. |
+| `enableFreeTier` | bool | `False` |  | Enable free tier of cosmos db. |
```
Comment (AlexanderSehr) [inline modules/Microsoft.DocumentDB/databaseAccounts/readme.md:42]: Was this change intended?

**Comment by AlexanderSehr** [comment]
Hey @samuelya thank you for your great contribution. Note that I added a few small comments. 
Aside from that I just wanted to give you a heads up that we're currently implementing a breaking change in the module folder paths as in #3165 and will merge your PR in after those are in.

File: modules/container-service/managed-clusters/main.bicep [hunk 4]
```diff
@@ -619,6 +631,20 @@ module managedCluster_roleAssignments '.bicep/nested_roleAssignments.bicep' = [f
   }
 }]
 
+resource dnsZone 'Microsoft.Network/dnsZones@2018-05-01' existing = if (dnsZoneResourceId != null && webApplicationRoutingEnabled) {
+  name: last(split(dnsZoneResourceId, '/'))!
+}
+
+resource dnsZone_roleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = if (dnsZoneResourceId != null && webApplicationRoutingEnabled) {
```
Comment (AlexanderSehr) [inline modules/container-service/managed-clusters/main.bicep:638]: Hey @samuelya, I wonder if it would make sense to add an additional condition (true/false - default: true) to control the role assignment with. Imagining a case where somebody wants to provide the DNS Zone but where the deploying SPN doesn't have `User Access Administrator` permissions I could imaging this could be a pitfall. 
I guess it is safe to say that is a general 'issue' to consider for CARML as we have other cases where we assign roles too. @eriqua any thoughts from your side too?

Comment (AlexanderSehr) [reply modules/container-service/managed-clusters/main.bicep:638]: PS: I am in favor of the role assignment and just want to make sure we do it the right way ;)

File: modules/container-service/managed-clusters/main.bicep [hunk 5]
```diff
@@ -416,6 +425,12 @@ resource managedCluster 'Microsoft.ContainerService/managedClusters@2022-11-01'
     agentPoolProfiles: primaryAgentPoolProfile
     linuxProfile: (empty(aksClusterSshPublicKey) ? null : aksClusterLinuxProfile)
     servicePrincipalProfile: (empty(aksServicePrincipalProfile) ? null : aksServicePrincipalProfile)
+    ingressProfile: {
+      webAppRouting: {
+        enabled: webApplicationRoutingEnabled
+        dnsZoneResourceId: !empty(dnsZoneResourceId) ? any(dnsZoneResourceId) : null
+      }
+    }
```
Comment (ahmadabdalla) [inline modules/container-service/managed-clusters/main.bicep:433]: @samuelya can you please provide some test parameters for this change so we can validate the workflow in case someone would like to use these parameters. We have a couple of AKS test deployments [here](https://github.com/Azure/ResourceModules/tree/main/modules/container-service/managed-clusters/.test). You can select either the Azure or the kubenet version. This requires you to create a private DNS zone in the dependencies which then can be referenced in the module deployment. Happy to support you in this if you need more guidance.

**Comment by AlexanderSehr** [comment]
@krbar is there any update on this?

File: modules/container-service/managed-clusters/.test/min/main.test.bicep [hunk 6]
```diff
@@ -17,9 +17,6 @@ param serviceShort string = 'csmmin'
 @description('Optional. Enable telemetry via a Globally Unique Identifier (GUID).')
 param enableDefaultTelemetry bool = true
 
-@description('Optional. A token to inject into the name of each resource.')
-param namePrefix string = '[[namePrefix]]'
```
[the remark this answers is not in the document]
Comment (AlexanderSehr) [reply modules/container-service/managed-clusters/.test/min/main.test.bicep]: Could you add the missing reference?
